control-plane/grpc/src/operations/volume/traits.rs

```diff
@@ -0,0 +1,908 @@
+                },
+                None => None,
+            },
+        );
+        Ok(target)
+    }
+}
+
+impl From<VolumeTarget> for volume::VolumeTarget {
+    fn from(target: VolumeTarget) -> Self {
+        volume::VolumeTarget {
+            node_id: target.node().to_string(),
+            nexus_id: Some(target.nexus().to_string()),
+            protocol: match target.protocol() {
+                None => None,
+                Some(protocol) => {
+                    let protocol: volume::VolumeShareProtocol = (*protocol).into();
+                    Some(protocol as i32)
+                }
+            },
+        }
+    }
+}
+
+impl From<volume::VolumeShareProtocol> for VolumeShareProtocol {
+    fn from(src: volume::VolumeShareProtocol) -> Self {
+        match src {
+            volume::VolumeShareProtocol::Nvmf => Self::Nvmf,
+            volume::VolumeShareProtocol::Iscsi => Self::Iscsi,
+        }
+    }
+}
+
+impl From<VolumeShareProtocol> for volume::VolumeShareProtocol {
+    fn from(src: VolumeShareProtocol) -> Self {
+        match src {
+            VolumeShareProtocol::Nvmf => Self::Nvmf,
+            VolumeShareProtocol::Iscsi => Self::Iscsi,
+        }
+    }
+}
+
+impl TryFrom<get_volumes_request::Filter> for Filter {
+    type Error = ReplyError;
+    fn try_from(filter: get_volumes_request::Filter) -> Result<Self, Self::Error> {
+        Ok(match filter {
+            get_volumes_request::Filter::Volume(volume_filter) => {
+                Filter::Volume(match VolumeId::try_from(volume_filter.volume_id) {
+                    Ok(volumeid) => volumeid,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                })
+            }
+        })
+    }
+}
+
+/// trait to be implemented for CreateVolume operation
+pub trait CreateVolumeInfo: Send + Sync {
+    /// uuid of the volume
+    fn uuid(&self) -> VolumeId;
+    /// size in bytes of the volume
+    fn size(&self) -> u64;
+    /// no of replicas of the volume
+    fn replicas(&self) -> u64;
+    /// volume policy of the volume, i.e self_heal
+    fn policy(&self) -> VolumePolicy;
+    /// topology configuration of the volume
+    fn topology(&self) -> Option<Topology>;
+    /// labels to be added to the volumes for topology based scheduling
+    fn labels(&self) -> Option<VolumeLabels>;
+}
+
+impl CreateVolumeInfo for CreateVolume {
+    fn uuid(&self) -> VolumeId {
+        self.uuid.clone()
+    }
+
+    fn size(&self) -> u64 {
+        self.size
+    }
+
+    fn replicas(&self) -> u64 {
+        self.replicas
+    }
+
+    fn policy(&self) -> VolumePolicy {
+        self.policy.clone()
+    }
+
+    fn topology(&self) -> Option<Topology> {
+        self.topology.clone()
+    }
+
+    fn labels(&self) -> Option<VolumeLabels> {
+        self.labels.clone()
+    }
+}
+
+impl CreateVolumeInfo for CreateVolumeRequest {
+    fn uuid(&self) -> VolumeId {
+        VolumeId::try_from(self.uuid.clone().unwrap()).unwrap()
```

Comment:
hmm, this is a bit annoying... How wonder if you need to have another CreateVolumeRequest that impl's a TryFrom and allows us to return errors? We have the same issue below on the TODO.
@paulyoong

Reply:
For volumeId it maybe fine as uuid is system generated, so there shouldn't be anycase ideally where we would be able to convert. But yeah, for topology I am not really sure what to do.

Reply:
Resolving this as we have handled this using the `ValidateRequestTypes` trait.